### PR TITLE
Remove dead code from same-shard decider

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today the same-shard allocation decider falls back to checking the
hostname if the node has no host address. In practice nodes will always
have an address so the fallback is dead code. This commit removes that
dead code.

Relates #80702 which will add the ability to distinguish nodes by
hostname regardless of whether they have an address or not, and #80767
which optimizes this area of code - this refactoring should make the
optimization simpler.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Good point, addressed in 19f9d00.

**Review by henningandersen — APPROVED** [review]
LGTM (once CI passes)

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java [hunk 1]
```diff
@@ -91,23 +91,12 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
...
-                        checkNodeOnSameHostAddress = true;
-                    }
-                } else if (Strings.hasLength(checkNode.node().getHostName()) && Strings.hasLength(node.node().getHostName())) {
-                    if (checkNode.node().getHostName().equals(node.node().getHostName())) {
-                        checkNodeOnSameHostName = true;
-                    }
-                }
-                if (checkNodeOnSameHostAddress || checkNodeOnSameHostName) {
+                assert Strings.hasLength(checkNode.node().getHostAddress()) : checkNode;
+                assert Strings.hasLength(node.node().getHostAddress()) : node;
```
Comment (idegtiarenko) [inline server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java:95]: NIT: `node.node().getHostAddress()` might be asserted once before the loop

**Review by idegtiarenko — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/SameShardAllocationDecider.java]: Good point, fixed in d3cbade5ae04939ad89d3d2104e0f00393db1df2.

Not really sure these asserts are useful outside the context of this change. I imagine we'll see them in a year or two and wonder why on earth they're there...

**Pull request merged**